### PR TITLE
core-math: provide missing FP16 host math overloads

### PR DESCRIPTION
**Pull request by maleadt**
When `cl_khr_fp16` is enabled, the host CPU kernel library advertises half support but several `_cl_*` half math entry points are absent in vectorized builtin configurations, because the SLEEF/libmvec wrappers only cover float and double. SPIR wrappers and OpenCL C kernels can then fail at program creation or link time with unresolved half math symbols.

Add the missing FP16 overloads to the host core-math library. Operations that have native LLVM half support stay in `additionalf16.cl`; the remaining builtins live in `promotedf16.cl` and are implemented by promoting through float, handling vector forms and pointer address spaces explicitly. Include the new source in both vectorized and non-vectorized host builds, and add the needed FP16 builtin declaration helpers.

Without this, the added tests fails on a build like:

```
CMAKE_GENERATOR=Ninja
CMAKE_BUILD_TYPE=Debug
ENABLE_TESTS=ON
ENABLE_SLEEF=ON
ENABLE_HOST_CPU_VECTORIZE_SVML=ON
ENABLE_HOST_CPU_VECTORIZE_LIBMVEC=OFF
ENABLE_HOST_CPU_VECTORIZE_SLEEF=OFF
ENABLE_HOST_CPU_VECTORIZE_BUILTINS=ON
KERNELLIB_HOST_CPU_VARIANTS=distro
LLVM_VERSION_FULL=20.1.8jl
LLVM_LIBS=-lLLVM-20jl
CPU_USE_LLD_LINK=defined
HOST_CPU_ENABLE_SPIRV=defined
OCL_KERNEL_TARGET_CPU="znver5"
HOST_LD_FLAGS="-shared -nostartfiles /opt/intel/oneapi/compiler/latest/lib/libsvml.a /opt/intel/oneapi/compiler/latest/lib/libirc.a"
```

... with:

```
Running test test_fp16_math_builtins...

Unknown OpenCL error 1 in clCreateProgram call failed
 on line 42
FAIL

Error(s) while linking:
Cannot find symbol _Z9_cl_rootnDv8_DhDv8_i in kernel library
Cannot find symbol _Z9_cl_rootnDhi in kernel library
Cannot find symbol _Z9_cl_ldexpDhi in kernel library
Cannot find symbol _Z10_cl_remquoDhDhPU9CLgenerici in kernel library
Cannot find symbol _Z8_cl_powrDv8_DhS_ in kernel library
Cannot find symbol _Z9_cl_ilogbDv8_Dh in kernel library
Cannot find symbol _Z13_cl_nextafterDv8_DhS_ in kernel library
Cannot find symbol _Z8_cl_modfDhPU9CLgenericDh in kernel library
Cannot find symbol _Z8_cl_logbDv8_Dh in kernel library
Cannot find symbol _Z8_cl_powrDhDh in kernel library
Cannot find symbol _Z8_cl_logbDh in kernel library
Cannot find symbol _Z10_cl_remquoDv8_DhS_PU9CLgenericDv8_i in kernel library
Cannot find symbol _Z8_cl_pownDhi in kernel library
Cannot find symbol _Z13_cl_remainderDv8_DhS_ in kernel library
Cannot find symbol _Z13_cl_nextafterDhDh in kernel library
Cannot find symbol _Z9_cl_ilogbDh in kernel library
Cannot find symbol _Z9_cl_ldexpDv8_DhDv8_i in kernel library
Cannot find symbol _Z8_cl_modfDv8_DhPU9CLgenericS_ in kernel library
Cannot find symbol _Z13_cl_remainderDhDh in kernel library
Cannot find symbol _Z8_cl_pownDv8_DhDv8_i in kernel library

Device cpu-skylake-avx512-AMD Ryzen 9 9950X 16-Core Processor failed to build the program
CL_BUILD_PROGRAM_FAILURE in clBuildProgram on line 686A
```

Reduced from an existing failure in OpenCL.jl

### CODE REVIEW

**Comment by maleadt** [comment]
OpenCL.jl failure are expected; these are broken tests now working. I'm backporting this change onto our build of PoCL so that I can update the tests and prevent the red mark as soon as this is merged.

**Comment by maleadt** [comment]
Merged in OpenCL.jl; CI is all green now here.

**Comment by franz** [comment]
I've done an experiment, this PR + applying this patch:

```patch
diff --git a/CMakeLists.txt b/CMakeLists.txt
index fc4c33795..403240c7c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1646,7 +1646,6 @@ endif()
 # LLVM 15 crashes on some code; LLVM 16 seems to work but fails with ldexp; LLVM 18 fails with isinf
 # CONFORMANCE=ON disables FP16 b/c PoCL support is incomplete
 if((LLVM_VERSION_MAJOR GREATER_EQUAL 19)
-   AND (NOT ENABLE_CONFORMANCE)
    AND (NOT I386)
    AND (CMAKE_HOST_SYSTEM_NAME STREQUAL "Linux")
    AND CLANG_SUPPORTS_FLOAT16_ON_CPU
```

then build PoCL with `cmake -DENABLE_CONFORMANCE=1 -DENABLE_TESTSUITES=conformance ~/source/pocl` 

then test the math builtins with `examples/conformance/src/conformance-build/test_conformance/math_brute_force/test_bruteforce -d -f -r -8` (this should test only fp16 math builtins, with vector width 8)

five tests failed:

```
frexp...
 30:           frexp fp16     ERROR: clBuildProgram failed! (CL_BUILD_PROGRAM_FAILURE from conformance/test_common/harness/kernelHelpers.cpp:813)
...
Build not successful for device "cpu-znver3-AMD Ryzen 9 5900X 12-Core Processor", status: CL_BUILD_ERROR
Build log for device "cpu-znver3-AMD Ryzen 9 5900X 12-Core Processor" is: ------------
Error(s) while linking:
Cannot find symbol _Z9_cl_frexpDv8_DhPU9CLgenericDv8_i in kernel library

lgamma_r...
 48:        lgamma_r fp16     ERROR: clBuildProgram failed! (CL_BUILD_PROGRAM_FAILURE from conformance/test_common/harness/kernelHelpers.cpp:813)
Build not successful for device "cpu-znver3-AMD Ryzen 9 5900X 12-Core Processor", status: CL_BUILD_ERROR
Build log for device "cpu-znver3-AMD Ryzen 9 5900X 12-Core Processor" is: ------------
Error(s) while linking:
Cannot find symbol _Z12_cl_lgamma_rDv8_DhPU9CLgenericDv8_i in kernel library

powr...
 62:            powr fp16     ..
ERROR: powr8: nan ulp error at {0x1p+0 (0x3c00), inf (0x7c00)}
Expected: nan  (half 0x7e00)
Actual: 0x1p+0 (half 0x3c00) at index: 224222
ERROR: ThreadPool_Do: TestHalf failed
! (CL_DEVICE_NOT_FOUND from /home/franz/0/build/b_pocl_main_22/examples/conformance/src/conformance/test_conformance/math_brute_force/binary_half.cpp:734)
powr FAILED

rsqrt...
 68:           rsqrt fp16
ERROR: rsqrt8: -1.196974 ulp error at 0x1.0ep-16 (half 0x010e)
Expected: 0x1.f28p+7 (half 0x5bca)
Actual: 0x1.f24p+7 (half 0x5bc9)
ERROR: ThreadPool_Do: TestHalf failed
! (CL_DEVICE_NOT_FOUND from /home/franz/0/build/b_pocl_main_22/examples/conformance/src/conformance/test_conformance/math_brute_force/unary_half.cpp:448)
rsqrt FAILED

sincos...
 71:          sincos fp16     ERROR: clBuildProgram failed! (CL_BUILD_PROGRAM_FAILURE from conformance/test_common/harness/kernelHelpers.cpp:813)
Build not successful for device "cpu-znver3-AMD Ryzen 9 5900X 12-Core Processor", status: CL_BUILD_ERROR
Build log for device "cpu-znver3-AMD Ryzen 9 5900X 12-Core Processor" is: ------------
Error(s) while linking:
Cannot find symbol _Z10_cl_sincosDv8_DhPU9CLgenericS_ in kernel library
```

AFAICT your PR doesn't touch `rsqrt`, `sincos` and `lgamma_r` so those are unrelated bugs (sincos & lgamma_r probably just missing some overloads); `powr` might just need to handle a few extra corner cases; and it seems that `frexp` is missing vector overloads. Otherwise LGTM.

**Comment by maleadt** [comment]
> AFAICT your PR doesn't touch `rsqrt`, `sincos` and `lgamma_r` so those are unrelated bugs (sincos & lgamma_r probably just missing some overloads); `powr` might just need to handle a few extra corner cases; and it seems that `frexp` is missing vector overloads.

I had Claude had a go at this. With conformance enabled 8/8 FP16 tests pass, but I couldn't test the bruteforce on the Windows box I'm now booted in.

**Comment by franz** [comment]
@maleadt thanks, I've rerun the bruteforce test and it looks good - only rsqrt is failing, but i have a fix for that already.

**Review by franz — APPROVED** [review]
LGTM
